lib/modules/manager/devcontainer/extract.spec.ts

```diff
@@ -0,0 +1,323 @@
+  }`),
+  malformedImage: getFixture(`
+  {
+    "image:": "devcontainer.registry.renovate.com/test/image:1.2.3"
+  }`),
+  noImageOrFeatures: getFixture('{}'),
+  nullFeatures: getFixture(`
+  {
+    "features": null
+  }`),
+  nullImage: getFixture(`
+  {
+    "image": null
+  }`),
+  nullImageAndFeatures: getFixture(`
+  {
+    "image": null,
+    "features": null
+  }`),
+  allKindsOfFeatures: getFixture(`
+  {
+    "features": {
+      "devcontainer.registry.renovate.com/test/feature:1.2.3": {},
+      "./localfeature": {},
+      "devcontainer.registry.renovate.com/test/feature/other.tgz": {}
+    }
+  }`),
+};
+
+describe('modules/manager/devcontainer/extract', () => {
+  describe('extractPackageFile()', () => {
+    afterEach(() => {
+      jest.restoreAllMocks();
+    });
+
+    it('returns null when file is empty', () => {
+      // Arrange
+      const content = '';
+      const packageFile = '';
+      const extractConfig = {};
+      // Act
+      const result = extractPackageFile(content, packageFile, extractConfig);
+
+      // Assert
+      expect(result).toBeNull();
+    });
+
+    it('returns null when file contents is malformed', () => {
+      // Arrange
+      const content = 'malformed json}}}';
+      const packageFile = '';
+      const extractConfig = {};
+      // Act
+      const result = extractPackageFile(content, packageFile, extractConfig);
+
+      // Assert
+      expect(result).toBeNull();
+    });
+
+    it('returns feature images when only features is defined', () => {
+      // Arrange
+      const content = fixtures.featuresOnly.content;
+      const packageFile = fixtures.featuresOnly.path;
+      const extractConfig = {};
+      // Act
+      const result = extractPackageFile(content, packageFile, extractConfig);
+
+      // Assert
+      expect(result).not.toBeNull();
+      expect(result?.deps).not.toBeNull();
+      expect(result?.deps.length).toBe(2);
+      expect(result?.deps).toMatchInlineSnapshot(`
+      [
+        {
+          "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
+          "currentDigest": undefined,
+          "currentValue": "1.2.3",
+          "datasource": "docker",
+          "depName": "devcontainer.registry.renovate.com/test/features/first",
+          "replaceString": "devcontainer.registry.renovate.com/test/features/first:1.2.3",
+        },
+        {
+          "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
+          "currentDigest": undefined,
+          "currentValue": "4.5.6",
+          "datasource": "docker",
+          "depName": "devcontainer.registry.renovate.com/test/features/second",
+          "replaceString": "devcontainer.registry.renovate.com/test/features/second:4.5.6",
+        },
+      ]
+      `);
+    });
+
+    it('returns image and feature images when both image and features are defined', () => {
```

Comment:
This is confusing to me because the message mixes two concepts:

- image as the field in the file?
- images as the thing you download  from Docker Hub?

Please try to rewrite this so it's clear which concept is meant in each part of the sentence.